library/std/src/fs.rs

```diff
@@ -615,6 +615,181 @@ impl File {
         self.inner.datasync()
     }
 
+    /// Acquire an exclusive advisory lock on the file. Blocks until the lock can be acquired.
+    ///
+    /// This acquires an exclusive advisory lock; no other file handle to this file may acquire
+    /// another lock.
+    ///
+    /// If this file handle already holds an advisory lock the exact behavior is
+    /// unspecified and platform dependent, including the possibility that it will deadlock.
+    /// However, if this method returns, then an exclusive lock is held.
+    ///
+    /// If the file not open for writing, it is unspecified whether this function returns an error.
+    ///
+    /// Note, this is an advisory lock meant to interact with [`lock_shared`], [`try_lock`],
+    /// [`try_lock_shared`], and [`unlock`]. Its interactions with other methods, such as [`read`]
+    /// and [`write`] are platform specific, and it may or may not cause non-lockholders to block.
```

Comment:
If it causes read and write for non-lockholders to block, it is a mandatory lock, not an advisory lock, right?

Reply:
It's platform specific whether or not it is mandatory (on Windows the lock is mandatory). Do you think it would be more clear to remove the word "advisory" from the API? I added it because that's the wording used by WASI, and it's frequently used when referring to locks on Unix

Reply:
Is it possible to use advisory locks on all platforms including Windows? I think having it be a mandatory lock, but only on some platforms will cause subtle issues that only show up on some platforms.

Reply:
Note that tools such as rustc and Cargo use either mandatory or advisory locks depending on the platform.

Reply:
A mandatory lock would prevent intentional modification of the file as well as I think intentional deletion by the end user, resulting in worse UX. For Cargo the file that is locked is supposed to remain empty and only the locking itself matters, so there the observable difference between advisory and mandatory locking is not all that big. For a general feature that will be used well beyond the specific case of Cargo, the UX differences and such matter a lot more though.

Reply:
> Is it possible to use advisory locks on all platforms including Windows? I think having it be a mandatory lock, but only on some platforms will cause subtle issues that only show up on some platforms.

No, Windows only supports mandatory locks and Unix only supports advisory locks. At least as far as I am aware.

In the ACP's comments I was told to document that the API may be advisory or mandatory. https://github.com/rust-lang/libs-team/issues/412#issuecomment-2246255810

Reply:
At least on modern Windows, you can delete an locked file.